### PR TITLE
Feature/arrows and more

### PR DESCRIPTION
**Pull request by ranjanbuilds**
Was trying to accomplish a few things: 

1) Got rid of baked in svgs for Arrows and now require user to pass in an arrow svg url to the component

2) Rather than calling the changeToActiveSlide() in many functions each time the activeIndex is changed, using an event handler to watch vm.activeIndex and trigger changeToActiveSlide(). 

Not using a setter here because we want to trigger the function after the property has changed. Catalyst for this change is so we can two-way bind activeSlideIndex and pass it in dynamically from the parent component as well.

3) Assigning unique classes to each slide based on their index number, i.e. slide0, slide1, etc. In the dissolve-transition related functions that Kevin introduced, we are relying heavily on the 'active' class being updated properly prior to certain functions firing. Now, instead we look at the activeSlideIndex in the vm and search for the relevant unique slide class - this bypasses any race-condition events.

### CODE REVIEW

File: src/donejs-carousel-plugin.js [hunk 1]
```diff
@@ -29,18 +30,32 @@ export const isMobile = function(isTablet) {
...
 	},
 	/**
-	* @property {number} lastSlideIndex index of last slide in the array
+	* @property {number} activeSlide 
+	* @description active slide dom element
+	*/
+	activeSlide: {
+		get() {
+			if (this.activeSlideIndex !== undefined && this.classSelector) {
+				return $(`${this.classSelector} .slide${this.activeSlideIndex}`);
```
Comment (kcarmonamurphy-tc) [inline src/donejs-carousel-plugin.js]: active slide can also be referenced by doing `$(``${this.classSelector} .slide.active``);` I think this would obviate the need for you to add the active slide index to the template?

File: src/donejs-carousel-plugin.js [hunk 2]
```diff
@@ -396,20 +419,22 @@ export const ViewModel = DefineMap.extend({
 	*
 	*/
 	swipeEnd() {
+		let touchEvent = this.defineTouchEvent(event);
```
Comment (kcarmonamurphy-tc) [inline src/donejs-carousel-plugin.js]: that's weird. wasn't this code always in the repo? not sure why it says it's being added here.

File: src/donejs-carousel-plugin.js [hunk 3]
```diff
@@ -504,26 +528,26 @@ export const ViewModel = DefineMap.extend({
...
 		let isLastSlide = this.activeSlideIndex == this.lastSlideIndex;
-
+		
 		// only fade the active slide if it's fading out into another slide, not on it's own
 		if (
 			(swipeAmount > 0 && !isFirstSlide) ||
 			(swipeAmount < 0 && !isLastSlide)
 		){
-			$(`${classSelector} .slide.active`).css({
+			this.activeSlide.css({
```
Comment (kevcom) [inline src/donejs-carousel-plugin.js]: good idea to make this a property.

File: src/donejs-carousel-plugin.js [hunk 4]
```diff
@@ -537,15 +561,14 @@ export const ViewModel = DefineMap.extend({
 	* active slide should have higher z-index value
 	*/
 	fadeToActiveSlide() {
-		let classSelector = this.classSelector;
-
+		
 		// only fade if slides are in carousel, not broken on desktop
 		if (!(this.isDesktop && this.carouselOptions.breakOnDesktop)) {
 			// transition length in ms
 			let transitionLength = 1000;
```
Comment (kevcom) [inline src/donejs-carousel-plugin.js]: should probably make this a `const`

File: src/donejs-carousel-plugin.stache [hunk 5]
```diff
@@ -30,7 +30,7 @@
 
 {{#if ./isDesktopBrowser}}
 	
-	{{#is ./slides.length 1}}
+	{{#if ./isLengthOneOrLess(./slides.length)}}
```
Comment (kevcom) [inline src/donejs-carousel-plugin.stache]: this means that now if there are no slides being passed in, the carousel div will still render but it'll be blank but probably still take up some space. before, it would be as if the carousel didn't exist. is this desired behaviour?

File: src/donejs-carousel-plugin.stache [hunk 6]
```diff
@@ -75,14 +75,16 @@
...
+						</svg>
 					</a>
 					
 					<a class="rightArrow arrow" 
 						on:click="./arrowClickHandler('right', scope.event)"
 					>
-						<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
-							<path d="M504 256C504 119 393 8 256 8S8 119 8 256s111 248 248 248 248-111 248-248zm-448 0c0-110.5 89.5-200 200-200s200 89.5 200 200-89.5 200-200 200S56 366.5 56 256zm72 20v-40c0-6.6 5.4-12 12-12h116v-67c0-10.7 12.9-16 20.5-8.5l99 99c4.7 4.7 4.7 12.3 0 17l-99 99c-7.6 7.6-20.5 2.2-20.5-8.5v-67H140c-6.6 0-12-5.4-12-12z"/>
+						<svg role="img" class="iconSvg" title="right nav arrow">
+							<use href="{{./carouselOptions.navArrows.rightSvgUrl}}"></use>
```
Comment (kevcom) [inline src/donejs-carousel-plugin.stache]: so with this can we specify an svg as a string or does it have to be a file?

File: src/test-and-demo-constants.js [hunk 7]
```diff
@@ -32,7 +32,6 @@ export const slideArray = [
 	{
 		slides: slides,
 		carouselOptions: {
-			navArrows: true,
```
Comment (kevcom) [inline src/test-and-demo-constants.js]: this should be updated to reflect new way of defining `navArrows`. along with most of the other ones in the file, except for where `navArrows` is set to `false`

Comment (ranjanbuilds) [reply src/test-and-demo-constants.js]: Not having navArrows defined is the same thing as setting to false. That's why just deleted it.

Comment (ranjanbuilds) [reply src/test-and-demo-constants.js]: Never mind, I'm an idiot, I did change the way arrows were feeding into the demo, but I forgot to push it from my laptop. Will do when I get home today.

File: src/donejs-carousel-plugin.js (hunk 1 above)
Comment (ranjanbuilds) [reply src/donejs-carousel-plugin.js]: yup, exactly

File: src/donejs-carousel-plugin.js (hunk 2 above)
Comment (ranjanbuilds) [reply src/donejs-carousel-plugin.js]: this code was in the swipeMove function, added it also in the swipeEnd because it should only be executing fully if dragging and touch conditions are met.

File: src/donejs-carousel-plugin.js (hunk 3 above)
Comment (ranjanbuilds) [reply src/donejs-carousel-plugin.js]: thanks

File: src/donejs-carousel-plugin.js (hunk 4 above)
Comment (kevcom) [reply src/donejs-carousel-plugin.js]: using `let` leaves the possibility for that variable to be changed, but we now that within the scope of this function `transitionLength` should never change, so better to use `const`.

But i'm being overly pedantic, it's not a huge deal.

File: src/donejs-carousel-plugin.js [hunk 8]
```diff
@@ -119,7 +119,7 @@ export const ViewModel = DefineMap.extend({
 	* @description passed in from the parent component 
 	*/
 	carouselOptions: {
-		type: 'observable',
+		type: 'any',
```
Comment (kevcom) [inline src/donejs-carousel-plugin.js]: what's the effect of doing this?

File: src/donejs-carousel-plugin.stache (hunk 5 above)
Comment (ranjanbuilds) [reply src/donejs-carousel-plugin.stache]: Before if no slides were passed in, it would render the full carousel (the else block). But I see your point. Actually this function isn't even necessary, because I am accounting for 0 slide scenario in the parent stache, displaying none on the carousel component if no slides and showing 'no slide' message. I'm going to revert it back to the previous version and assume the user will account for 0 slide scenario (for now).

File: src/demo/demo.stache [hunk 9]
```diff
@@ -16,13 +24,16 @@
 	<can-import from="@tucows/donejs-carousel-plugin/demo/demo.less!" />
 	<can-import from="@tucows/donejs-carousel-plugin" />
 	<tucows-donejs-carousel
-		activeSlideIndex:to='./activeSlideIndex'
+		activeSlideIndex:bind='./activeSlideIndex'
```
Comment (kevcom) [inline src/demo/demo.stache]: also just out of my own curiosity why the change here? what does this accomplish?

File: src/donejs-carousel-plugin.stache (hunk 6 above)
Comment (ranjanbuilds) [reply src/donejs-carousel-plugin.stache]: It's got to be a in a file. Can revisit this at some point if you think that's too stringent.

File: src/demo/demo.stache (hunk 9 above)
Comment (ranjanbuilds) [reply src/demo/demo.stache]: In the TV channel component when we use the scroller to flip between channel types, I need to tell the carousel to go back to slide 0. This lets me do that.

File: src/donejs-carousel-plugin.js (hunk 4 above)
Comment (ranjanbuilds) [reply src/donejs-carousel-plugin.js]: cool, I'll change it - pedantic is good when it comes to code

File: src/test-and-demo-constants.js (hunk 7 above)
Comment (ranjanbuilds) [reply src/test-and-demo-constants.js]: I made the commit but it isn't showing up on this pull request. I'll close it and reopen.

File: src/donejs-carousel-plugin.js [hunk 10]
```diff
@@ -142,26 +165,30 @@ export const ViewModel = DefineMap.extend({
...
+	* @description whether or not the current environment is browser
 	*/
 	isDesktopBrowser: {
 		type: 'boolean',
 		value: platform.isDesktopBrowser
 	},
 	/**
-	* @property {boolean} classSelector specifies the appropriate carousel class especially if there are multiple carousels on page
+	* @property {boolean} classSelector 
+	* @description specifies the appropriate carousel class especially if there are multiple carousels on page
```
Comment (kevcom) [inline src/donejs-carousel-plugin.js]: good work fixing this up to conform to documentjs guidelines!

File: src/donejs-carousel-plugin.js (hunk 8 above)
Comment (ranjanbuilds) [reply src/donejs-carousel-plugin.js]: let's the vm know that live binding isn't necessary for this property. Which it isn't, because it's not going to be changing after the initial setup.

File: src/donejs-carousel-plugin.stache (hunk 5 above)
Comment (kevcom) [reply src/donejs-carousel-plugin.stache]: alright cool 👍
